### PR TITLE
chore: add http2 support

### LINKED ISSUE #27: http2 compat tests
Maybe we should have some tests against http2 compat?

### PR DESCRIPTION
**Pull request by bjohansebas**
With this, we officially add support for HTTP/2, and it would fix any previous issues with HTTP/2 and unblock middleware and Express itself from supporting HTTP/2, if the failure was due to this module.

Important changes:

* The entire test suite is refactored to have:
  * `test/general.js`: tests run against both HTTP/1 and HTTP/2 for shared cases that don’t require version-specific changes.
  * `test/http1.js`: HTTP/1-specific tests that couldn’t be shared in `test/general.js`.
  * `test/http2.js`: HTTP/2-specific tests, covering both Node.js’s compatibility layer and the native implementation without the compatibility layer (extracted from #100, thanks @GroophyLifefor)
  
### Important things to keep in mind with HTTP/2

- HTTP/2 does not have the `upgrade` event by specification; see RFC 9113, section 8.6
- HTTP/2 and Node.js’s implementation handle the `connect` event a bit differently. Unlike HTTP/1, it is not finalized immediately, and we don’t have the `upgrade` property that HTTP/1 had (which indicated at the moment of the `connect` event that it was already finalized). Here, it is finalized only once the request has read all the data and there’s nothing left to read, as `IncomingMessage` normally does.
- For `Http2ServerResponse`, which is the equivalent of `ServerResponse` in HTTP/1, we only consider the `closed` property, which will be `true` if the connection has already been closed. In testing, we could also check if the `destroyed` property is `true`, but in our tests both always had the same value, and using `closed` seems more appropriate.
- For `Http2ServerRequest`, which is the equivalent of `IncomingMessage` in HTTP/1, we similarly check if the `closed` property is `true`, and we also verify whether it’s no longer receiving data and has read everything it needed to (see https://github.com/jshttp/on-finished/pull/87#discussion_r2681336172).

closes #100 closes #27

### CODE REVIEW

**Comment by jonchurch** [comment]
There's a race condition failing 5 tests. The request is racing the client destroy, so sometimes you'll get `ERR_HTTP2_STREAM_CANCEL` emitted on the client, and the server will never receive the request. aka the request is aborted before it reaches the server

Moving the destroy into the request handler will ensure they dont race each other:
```js
    it('should fire with error', function (done) {
      var client
      var server = http.createServer(function (req, res) {
        onFinished(res, function (_err) {
          server.close(done)
        })
        // destroy here, inside the handler
        client.destroy()
      })
```

File: test/http2.js [hunk 1]
```diff
@@ -0,0 +1,893 @@
...
+describe('isFinished(req)', function () {
+  it.skip('should be true after request finishes', function (done) {
+    var server = http.createServer(function (req, res) {
+      onFinished(req, function (err) {
+        assert.ifError(err)
+        assert.ok(onFinished.isFinished(req))
+        done()
+      })
+
+      // WHY?
```
Comment (jonchurch) [inline test/http2.js]: The current `isFinished` implementation in the PR doesnt account for the duplex nature of http2
(fix in https://github.com/jshttp/on-finished/pull/87#discussion_r2681352062)

Checking `stream.closed` represents the entire duplex stream, read and write sides. When the request has finished, the readable side is closed but the writable side (response) may still be open. `stream.closed` will only be true when both sides are done. When checking responses we can rely on `stream.closed` because if the stream is closed the response was either aborted or fully sent.

For http2 requests we need to check if `complete && !readable` to know that all data has been received and there's nothing else to read (or, as I found out in https://github.com/jshttp/on-finished/pull/87#discussion_r2681420697, we can check if `!socket.readable` but that feels cursed to me, like a leaky implementation detail)

this tests captures that I was wrong in https://github.com/jshttp/on-finished/pull/100#discussion_r2681145198 

which is very handy!

File: index.js
Comment (jonchurch) [inline index.js:96]: This should pass the WHY test for responses https://github.com/jshttp/on-finished/pull/87/changes#r2681336172

```js
function isFinished (msg) {
  const socket = msg.socket
  const stream = msg.stream
  const isHttp2 = stream && typeof stream.closed === 'boolean'

  // OutgoingMessage or Http2ServerResponse
  if (typeof msg.writableEnded === 'boolean') {
    if (isHttp2) return stream.closed
    return Boolean(msg.writableEnded || (socket && !socket.writable))
  }

  // IncomingMessage or Http2ServerRequest
  if (typeof msg.complete === 'boolean') {
    if (isHttp2) return Boolean(stream.closed || (msg.complete && !msg.readable))
    return Boolean(msg.upgrade || !socket || !socket.readable || (msg.complete && !msg.readable))
  }

  // don't know
  return undefined
}
```

Comment (jonchurch) [reply index.js:96]: Or, less readable but still correct based on running these tests, we can keep the IncomingMessage / Http2ServerRequest checking the same as master:

```js
function isFinished (msg) {
  const socket = msg.socket

  if (typeof msg.writableEnded === 'boolean') {
    // OutgoingMessage or Http2ServerResponse
    const stream = msg.stream
    if (stream && typeof stream.closed === 'boolean') {
      // Http2ServerResponse
      return stream.closed
    }
    return Boolean(msg.writableEnded || (socket && !socket.writable))
  }

  if (typeof msg.complete === 'boolean') {
    // IncomingMessage or Http2ServerRequest
    return Boolean(msg.upgrade || !socket || !socket.readable || (msg.complete && !msg.readable))
  }

  // don't know
  return undefined
}
```

When the request ends the socket stops being readable, so h2 will resolve true for `!socket.readable`. Which feels weird, bc h1 needs to check `msg.complete && !msg.readable`. The difference seems to be that the socket in h1 is the connection, and in h2 its a proxy for the stream. h2's response socket becomes undefined when it is finished (!)

The first approac above is definitely the more intuitive

**Comment by bjohansebas** [comment]
Okay, I only have two tests left to finish, bring over the tests from #100, add some documentation, and update the PR description.

**Comment by bjohansebas** [comment]
I’ve updated the PR description and brought in the changes from #100 (squashed, to include only the tests).

File: index.js
Comment (bjohansebas) [reply index.js:96]: I always learn something new from you whenever you do a review.

**Comment by Phillip9587** [comment]
Nice work @bjohansebas 🎉

**Comment by bjohansebas** [comment]
There will probably be edge cases, but once the version is released they could be quickly caught in the middleware and in Express itself, since webpack-dev-server no longer uses SPDY, and we’ll know how HTTP/2 behaves in Express with that package.

**Review by Phillip9587 — APPROVED** [review]
LGTM

**Comment by Phillip9587** [comment]
@bjohansebas Maybe add a entry to the `HISTORY.md` but we can do that later during the release.

**Comment by bjohansebas** [comment]
Okay, I think I found another edge case. I’m testing this with webpack-dev-server, and it looks like the failure comes from here, or possibly from finalhandler, I’m not sure, but I’m pretty confident it originates here. I don’t have a standalone example yet, since I’m testing it directly through webpack-dev-server tests.

edit: Okay, no, I didn’t find an edge case. I think it’s more about how properties are inherited. The thing is that by the time it reaches on-finished, neither the stream property nor the socket property exists. It’s weird and seems to be related to Express.

Confirmed: it’s Express. In https://github.com/expressjs/express/blob/a479419b16f5b97eb20f5dbae5848708ff30ce2d/lib/application.js#L169, setPrototypeOf causes neither stream nor socket to be accessibles

**Review by UlisesGascon — APPROVED** [review]
Great effort @bjohansebas! :clap: :clap: :clap:

**Comment by UlisesGascon** [comment]
are we sure that this is considered as major and not minor? based on the changes on the `index.js` IMO is minor

**Comment by bjohansebas** [comment]
It’s more of a minor change, the major change was in https://github.com/jshttp/on-finished/pull/56. The logic of this pr could even be backported to the current line (HTTP/2 won’t work perfectly, but for isFinished it will, while for onFinished we won’t know)

**Review by efekrskl — APPROVED** [review]

**Review by GroophyLifefor — APPROVED** [review]

**Comment by Phillip9587** [comment]
Thanks for all the hard work @bjohansebas 🎉
